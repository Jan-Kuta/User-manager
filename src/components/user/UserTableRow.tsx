import React, {ReactElement, useState} from 'react';
import Tooltip from '@material-ui/core/Tooltip';
import TableCell from '@material-ui/core/TableCell';
import TableRow from '@material-ui/core/TableRow';
import IconButton from '@material-ui/core/IconButton';
import Collapse from '@material-ui/core/Collapse';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Hidden from '@material-ui/core/Hidden';
import makeStyles from "@material-ui/core/styles/makeStyles";
import KeyboardArrowUpIcon from "@material-ui/icons/KeyboardArrowUp";
import KeyboardArrowDownIcon from "@material-ui/icons/KeyboardArrowDown";
import DeleteIcon from "@material-ui/icons/Delete";
import EditIcon from "@material-ui/icons/Edit";
import NewIcon from "@material-ui/icons/FiberNew";
import StarIcon from "@material-ui/icons/Star";
import ValidatedIcon from "@material-ui/icons/CheckCircle";
import { green, yellow, orange } from '@material-ui/core/colors';
import {User, UserState} from "../../types/userTypes";
import {useTranslation} from "react-i18next";
import {UserDetail} from "./UserDetail";


type Props = {
    user: User
    openUserEditDialog: (user: User) => void
    openUserDeleteDialog: (user: User) => void
}

const useStyles = makeStyles((theme) => ({
    root: {
        '& > *': {
            borderBottom: 'unset',
        },
    },
    nowrap: {
        whiteSpace: 'nowrap',
    },
    marginLeft: {
        marginLeft: theme.spacing(2),
    },
    flexBetween: {
        display: 'flex',
        justifyContent: 'space-between',
        marginTop: theme.spacing(2),
    },
    detailRoot: {
        backgroundColor: theme.palette.background.default
    }
}));

export const UserTableRow = (props: Props) => {
    const { user, openUserEditDialog, openUserDeleteDialog } = props;
    const { t } = useTranslation();
    const [openDetail, setOpenDetail] = useState(false);
    const classes = useStyles();

    const getStateIcon = (state: UserState) => {
        const wrapWithTooltip = (icon: ReactElement) => <Tooltip placement="left" title={t(state).toString()}>{icon}</Tooltip>;

        if (state === UserState.NEW) {
            return wrapWithTooltip(<NewIcon style={{ color: orange[500] }} />);
        }

        if (state === UserState.VIP) {
            return wrapWithTooltip(<StarIcon style={{ color: yellow[500] }} />);
        }

        return wrapWithTooltip(<ValidatedIcon style={{ color: green[500] }} />);
    };

    const handleOpenEditForm = () => {
        openUserEditDialog(user);
    };

    const handleOpenDeleteDialog = () => {
        openUserDeleteDialog(user);
    };

    return (
        <React.Fragment>
            <TableRow className={classes.root}>
                <TableCell>
                    <IconButton aria-label="expand row" size="small" onClick={() => setOpenDetail(!openDetail)}>
                        {openDetail ? <KeyboardArrowUpIcon /> : <KeyboardArrowDownIcon />}
                    </IconButton>
                </TableCell>
                <TableCell align="right">{user.firstName}</TableCell>
                <TableCell align="right">{user.lastName}</TableCell>
                <Hidden xsDown><TableCell align="right">{user.username}</TableCell></Hidden>
                <TableCell align="right">{getStateIcon(user.state)}</TableCell>
                <Hidden xsDown>
                    <TableCell className={classes.nowrap} align="right">
                        <Tooltip placement="left" title={t('Edit').toString()}>
                            <IconButton size="small" onClick={handleOpenEditForm}>
                                <EditIcon />
                            </IconButton>
                        </Tooltip>
                        <Tooltip placement="left" title={t('Delete').toString()}>
                            <IconButton className={classes.marginLeft} size="small" onClick={handleOpenDeleteDialog}>
                                <DeleteIcon />
                            </IconButton>
                        </Tooltip>
                    </TableCell>
                </Hidden>
            </TableRow>
            <TableRow className={classes.detailRoot}>
                <TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={6}>
                    <Collapse in={openDetail} timeout="auto" unmountOnExit>
                        <Box margin={1}>
                            <UserDetail user={user}/>
                            <Hidden smUp>
                                <div className={classes.flexBetween}>
                                    <Button
                                        size="small"
                                        variant="contained"
                                        aria-label={t('Edit')}
                                        color="primary"
                                        startIcon={<EditIcon />}
                                        onClick={handleOpenEditForm}
                                    >
                                        {t('Edit')}
                                    </Button>
                                    <Button
                                        size="small"
                                        variant="contained"
                                        aria-label={t('Delete')}
                                        color="secondary"
                                        startIcon={<DeleteIcon />}
                                        onClick={handleOpenDeleteDialog}
                                    >
                                        {t('Delete')}
                                    </Button>
                                </div>
                            </Hidden>
                        </Box>
                    </Collapse>
                </TableCell>
            </TableRow>
        </React.Fragment>
    );
};
