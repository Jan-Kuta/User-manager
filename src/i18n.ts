import i18n from "i18next";
import { initReactI18next } from "react-i18next";

// the translations
// (tip move them in a JSON file and import them)
const resources = {
    en: {
        translation: {
            "Language": "Language",
            "Note list": "Note list",
            "New note": "New note",
            "Note detail": "Note detail",
            "Title": "Title",
            "Delete": "Delete",
            "Edit": "Edit",
            "Dialog title": "Do you realy want to delete this item?",
            "Dialog description": "It can't be undone.",
            "Cancel": "Cancel",
            "OK": "OK",
            "Note edit": "Note edit",
            "Submit": "Submit",
            "User list": "User list",
            "User detail": "User detail",
            "Create user": "New user",
            "STATE_NEW": "New",
            "STATE_VALIDATED": "Validated",
            "STATE_VIP": "VIP",
            "Add user": "Add user",
            "Firstname": "Firstname",
            "Lastname": "Lastname",
            "Username": "Username",
            "State": "State",
            "Actions": "Actions"
        }
    },
    cs: {
        translation: {
            "Language": "Jazyk",
            "Note list": "Poznámky",
            "New note": "Nová poznámka",
            "Note detail": "Detail poznámky",
            "Title": "Popisek",
            "Delete": "Smazat",
            "Edit": "Upravit",
            "Dialog title": "Opravdu si přejete smazat tuto položku?",
            "Dialog description": "Tato akce nelze vrátit zpět.",
            "Cancel": "Zrušit",
            "OK": "Potvrdit",
            "Note edit": "Úprava poznámky",
            "Submit": "Uložit",
            "User list": "Seznam uživatelů",
            "User detail": "Další údaje",
            "Create user": "Nový uživatel",
            "Edit user": "Editace uživatele",
            "STATE_NEW": "Nový",
            "STATE_VALIDATED": "Ověřený",
            "STATE_VIP": "VIP",
            "Add user": "Přidat uživatele",
            "Basic info": "Základní údaje",
            "Firstname": "Jméno",
            "Lastname": "Příjmení",
            "Username": "Uživatelské jméno",
            "State": "Stav",
            "Actions": "Akce",
            "Email": "E-mail",
            "Phone": "Telefonní číslo",
            "Street": "Ulice",
            "City": "Město",
            "Zip": "Směrovací číslo",
            "Permanent address": "Adresa",
            "Contact address": "Kontaktní adresa",
            "Field is required": "Toto pole je povinné",
            "Addresses differ": "Zadat jinou kontaktní adresu",
            "Back": "Předchozí",
            "Next": "Následující",
            "Rows per page": "Počet záznamů na stránce",
            "Of": "z",
            "Delete confirmation": "Opravdu si přejete záznam smazat?",
            "Delete confirmation text": "Uživatel {{fullName}} bude smazán trvale. Akce nemůže být vrácen zpět.",
        }
    }
};

i18n
    .use(initReactI18next) // passes i18n down to react-i18next
    .init({
        resources,

        fallbackLng: ['en', 'cs'],

        lng: 'cs',

        keySeparator: false, // we do not use keys in form messages.welcome

        interpolation: {
            escapeValue: false // react already safes from xss
        }
    });

export default i18n;
