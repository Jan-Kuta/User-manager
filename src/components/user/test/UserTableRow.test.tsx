import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import {UserTableRow} from '../UserTableRow';
import {userList} from "./MockApplicationState";
import {initReactI18next} from "react-i18next";
import i18n from '../../../i18n';

describe("User table row", () => {
    initReactI18next.init(i18n);

    it("it should match", () => {
        const renderer = ShallowRenderer.createRenderer();

        renderer.render(<UserTableRow
            user={userList[0]}
            openUserDeleteDialog={jest.fn}
            openUserEditDialog={jest.fn}
        />);
        const result = renderer.getRenderOutput();

        expect(result).toMatchSnapshot();
    });
});
