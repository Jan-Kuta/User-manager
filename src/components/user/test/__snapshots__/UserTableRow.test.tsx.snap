// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User table row it should match 1`] = `
<React.Fragment>
  <ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)
        aria-label="expand row"
        onClick={[Function]}
        size="small"
      >
        <UNDEFINED />
      </ForwardRef(WithStyles)>
    </ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      align="right"
    >
      Jan
    </ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      align="right"
    >
      Kuta
    </ForwardRef(WithStyles)>
    <Hidden
      xsDown={true}
    >
      <ForwardRef(WithStyles)
        align="right"
      >
        Kuticzech
      </ForwardRef(WithStyles)>
    </Hidden>
    <ForwardRef(WithStyles)
      align="right"
    >
      <ForwardRef(WithStyles)
        placement="left"
        title="Nový"
      >
        <UNDEFINED
          style={
            Object {
              "color": "#ff9800",
            }
          }
        />
      </ForwardRef(WithStyles)>
    </ForwardRef(WithStyles)>
    <Hidden
      xsDown={true}
    >
      <ForwardRef(WithStyles)
        align="right"
      >
        <ForwardRef(WithStyles)
          placement="left"
          title="Upravit"
        >
          <ForwardRef(WithStyles)
            onClick={[Function]}
            size="small"
          >
            <UNDEFINED />
          </ForwardRef(WithStyles)>
        </ForwardRef(WithStyles)>
        <ForwardRef(WithStyles)
          placement="left"
          title="Smazat"
        >
          <ForwardRef(WithStyles)
            onClick={[Function]}
            size="small"
          >
            <UNDEFINED />
          </ForwardRef(WithStyles)>
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
    </Hidden>
  </ForwardRef(WithStyles)>
  <ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      colSpan={6}
      style={
        Object {
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <ForwardRef(WithStyles)
        in={false}
        timeout="auto"
        unmountOnExit={true}
      >
        <ForwardRef(StyledComponent)
          margin={1}
        >
          <UserDetail
            user={
              Object {
                "contactAddress": Object {
                  "city": "Nejdek",
                  "street": "Kollárova 964",
                  "zip": "362 21",
                },
                "email": "jan.kuta@email.cz",
                "firstName": "Jan",
                "id": "u1",
                "lastName": "Kuta",
                "permanentAddress": Object {
                  "city": "Ústí nad Labem",
                  "street": "Žežická 39",
                  "zip": "400 07",
                },
                "phone": "+420 775 246 369",
                "state": "STATE_NEW",
                "username": "Kuticzech",
              }
            }
          />
          <Hidden
            smUp={true}
          >
            <div>
              <ForwardRef(WithStyles)
                aria-label="Upravit"
                color="primary"
                onClick={[Function]}
                size="small"
                startIcon={<UNDEFINED />}
                variant="contained"
              >
                Upravit
              </ForwardRef(WithStyles)>
              <ForwardRef(WithStyles)
                aria-label="Smazat"
                color="secondary"
                onClick={[Function]}
                size="small"
                startIcon={<UNDEFINED />}
                variant="contained"
              >
                Smazat
              </ForwardRef(WithStyles)>
            </div>
          </Hidden>
        </ForwardRef(StyledComponent)>
      </ForwardRef(WithStyles)>
    </ForwardRef(WithStyles)>
  </ForwardRef(WithStyles)>
</React.Fragment>
`;
